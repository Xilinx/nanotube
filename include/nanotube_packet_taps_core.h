/**************************************************************************\
*//*! \file nanotube_packet_taps_core.h
** \author  Neil Turton <neilt@amd.com>
**  \brief  Interface for packet tap core logic functions.
**   \date  2020-09-23
*//*
\**************************************************************************/

/**************************************************************************
** Copyright (C) 2023, Advanced Micro Devices, Inc. All rights reserved.
** SPDX-License-Identifier: MIT
**************************************************************************/

#ifndef NANOTUBE_PACKET_TAPS_CORE_H
#define NANOTUBE_PACKET_TAPS_CORE_H

#include "nanotube_packet_taps.h"

#ifdef __cplusplus
extern "C" {
#endif

#include <stdint.h>

///////////////////////////////////////////////////////////////////////////

/*! Copy data between two buffers while rotating the order of the
**  bytes.
**
** \param buffer_out_length  The number of bytes in the output buffer.
** This must be a constant expression for synthesis.
**
** \param rot_buffer_length The number of bytes to rotate.  This must
** be at least as large as the input buffer length to avoid the input
** being truncated.  This must be a constant expression for synthesis.
**
** \param buffer_in_length  The number of bytes in the input buffer.
** This must be a constant expression for synthesis.
**
** \param rot_amount_bits  The number of bits required to specify
** the rotation amount.  This must be a constant expression
** for synthesis.
**
** \param buffer_out  The output buffer.
**
** \param buffer_in   The input buffer.
**
** \param rot_amount  The rotation amount as a number of bytes.
**
** Reads buffer_in_length bytes from buffer_in, pads them up to
** rot_buffer_length, rotates them by rot_amount bytes and writes the
** first buffer_out_length of them to buffer_out.
**
** The value written to rot_buffer[i] is:
**   rot_buffer[i] = (i >= buffer_in_length ? 0 : buffer_in[i])
**
** The rotate amount used is:
**    rot_amount_masked = ((1 << rot_amount_bits)-1) & rot_amount
**
** The value written to buffer_out[i] is:
**   rot_buffer[(i+rot_amount_masked)%rot_buffer_length].
**
** The rotate amount can be any value, but only the rot_amount_bits
** least significant bits are used.
*/
void nanotube_rotate_down(
  /* Constant parameters. */
  uint16_t buffer_out_length,
  uint16_t rot_buffer_length,
  uint16_t buffer_in_length,
  uint8_t rot_amount_bits,

  /* Outputs. */
  uint8_t *buffer_out,

  /* Inputs. */
  const uint8_t *buffer_in,
  uint16_t rot_amount);

///////////////////////////////////////////////////////////////////////////

/*! Copy bytes between buffers while shifting.
**
** \param buffer_out_length The number of bytes in the output buffer.
**
** \param buffer_out The output buffer.
**
** \param buffer_in The input buffer.
**
** \param shift_amount The shift amount in bits.
**
** Copy buffer_out_length bytes into buffer_out by shifting the bytes
** from buffer_in right and down in index.  The number of bytes in
** buffer_in is buffer_out_length+1 since some bits from input byte
** buffer_out_length might be shifted down into output byte
** buffer_out_length-1.  The maximum shift amount is 7.
**
*/
void nanotube_shift_down_bits(uint32_t buffer_out_length,
                              uint8_t *buffer_out,
                              const uint8_t *buffer_in,
                              uint8_t shift_amount);

///////////////////////////////////////////////////////////////////////////

/*! Pad a bit vector and concatenate it with itself.
**
** \param dup_bits_out The output vector.
**
** \param bit_vec_in The input vector.
**
** \param input_bit_length The number of bits in the input vector.
**
** \param padded_bit_length The number of bits in the padded vector.
**
** Produces a vector which includes two copies of the input vector,
** padded with zeros to the end of the byte.  The number of bytes in
** the output vector is the minimum reqired to hold 2*input_bit_length
** bits.
*/
void nanotube_duplicate_bits(uint8_t *dup_bits_out,
                             const uint8_t *bit_vec_in,
                             uint32_t input_bit_length,
                             uint32_t padded_bit_length);

///////////////////////////////////////////////////////////////////////////

/*! Determine the length of the packet, upto the specified maximum.
**
** \param resp_out The response structure, written by the tap with the
** result of the operation.
**
** \param state_inout An opaque state structure which should be
** initialised to nanotube_tap_packet_length_state_init and then
** passed to every call of the length tap.
**
** \param packet_word_eop A flag which indicates that the end of
** packet has arrived.
**
** \param packet_word_length The number of bytes in this packet word.
**
** \param req_in The request which is provided to the length tap.
**
** The packet length tap determines the length of the packet, bounding
** the result up to a specified parameter.  The request valid bit
** should be set as soon as the required maximum length is known and
** should remain set until the end of the packet.  If the request
** valid bit is never set, the full packet length will be provided at
** the end of the packet.
**
** The response valid bit will be set for exactly one word of the
** packet.  If the request valid bit is ever set for a packet then the
** response valid bit will be set no sooner than the request valid
** bit.  If the request valid bit is not set for a packet then the
** response valid bit will be set during the call for the end of
** packet word.
*/

void nanotube_tap_packet_length_core(
  /* Outputs. */
  struct nanotube_tap_packet_length_resp *resp_out,

  /* State. */
  struct nanotube_tap_packet_length_state *state_inout,

  /* Inputs. */
  bool packet_word_eop,
  uint16_t packet_word_length,
  const struct nanotube_tap_packet_length_req *req_in);


///////////////////////////////////////////////////////////////////////////

/*! Read a series of bytes from a packet.
**
** \param result_buffer_length  The size of the result buffer in bytes.
** This must be a constant expression for synthesis.
**
** \param result_buffer_index_bits The number of bits required for an
** index into the result buffer.  This must be a constant expression
** for synthesis.
**
** \param packet_buffer_length  The size of the packet buffer in bytes.
** This must be a constant expression for synthesis.
**
** \param packet_buffer_index_bits  The number of bits required for an
** index into the packet buffer.  This must be a constant expression
** for synthesis.
**
** \param resp_out The response structure, written by the tap with
** information about the read operation.
**
** \param result_buffer_inout  The result buffer, written by the
** tap with the bytes which were read.  The buffer is written
** incrementally, so the same buffer should be supplied to every call.
**
** \param state_inout  An opaque state structure which should be
** initialised to nanotube_tap_packet_read_state_init and then passed
** to every call to the read tap.
**
** \param packet_buffer_in  The input packet buffer.  Each call must
** supply packet_buffer_length packet bytes except the final call for
** a packet (with packet_word_eop set) which can be shorter.
**
** \param packet_word_eop Indicates whether this is the last call for
** a packet.
**
** \param req_in A non-null pointer to the read request.  The request
** is only acted on if the "valid" member is set.  After "valid" has
** been set, the request structure must remain constant until after
** the end of packet (the next call with packet_word_eop set).  The
** "valid" member must be set in the call for the word which contains
** the first byte to be read.
*/
void nanotube_tap_packet_read_core(
  /* Constant parameters */
  uint16_t result_buffer_length,
  uint8_t result_buffer_index_bits,
  uint16_t packet_buffer_length,
  uint8_t packet_buffer_index_bits,

  /* Outputs. */
  struct nanotube_tap_packet_read_resp *resp_out,
  uint8_t *result_buffer_inout,

  /* State. */
  struct nanotube_tap_packet_read_state *state_inout,

  /* Inputs. */
  const uint8_t *packet_buffer_in,
  bool packet_word_eop,
  uint16_t packet_word_length,
  const struct nanotube_tap_packet_read_req *req_in);


///////////////////////////////////////////////////////////////////////////

/*! Write a series of bytes to a packet with masking.
**
** \param request_buffer_length The size of the request buffer in
** bytes.  This must be a constant expression for synthesis.
**
** \param request_buffer_index_bits The number of bits required for an
** index into the request buffer.  This must be a constant expression
** for synthesis.
**
** \param packet_buffer_length The size of the packet buffer in bytes.
** This must be a constant expression for synthesis.
**
** \param packet_buffer_index_bits The number of bits required for an
** index into the packet buffer.  This must be a constant expression
** for synthesis.
**
** \param packet_buffer_out The output packet buffer.  The packet
** length is not modified, so the end of packet indicators are the
** same as for the input packet.
**
** \param state_inout An opaque state structure which should be
** initialised to nanotube_tap_packet_write_state_init and then passed
** to every call to the write tap.
**
** \param packet_buffer_in The input packet buffer.  Each call must
** supply packet_buffer_length packet bytes except the final call for
** a packet (with packet_word_eop set) which can be shorter.
**
** \param packet_word_eop Indicates whether this is the last call for
** a packet.
**
** \param packet_word_length The number of valid bytes in the input
** packet buffer.  This must be equal to packet_buffer_length except
** when packet_word_eop is set.
**
** \param req_in A non-null pointer to the write request.  The request
** is only acted on if the "valid" member is set.  After "valid" has
** been set, the request structure must remain constant until after
** the end of packet (the next call with packet_word_eop set).  The
** "valid" member must be set in the call for the word which contains
** the first byte to be written.
**
** \param request_bytes_in A non-null pointer to the bytes to be
** written to the packet.  The maximum number of bytes is given in the
** request structure.
**
** \param request_mask_in A non-null pointer to the byte enable bits.
** Each set bit in the mask indicates a byte to be written.  Each mask
** byte except the last refers to 8 data bytes with the least
** significant bit referring to the byte with the lowest offset.
*/
void nanotube_tap_packet_write_core(
  /* Constant parameters */
  uint16_t request_buffer_length,
  uint8_t request_buffer_index_bits,
  uint16_t packet_buffer_length,
  uint8_t packet_buffer_index_bits,

  /* Outputs. */
  uint8_t *packet_buffer_out,

  /* State. */
  struct nanotube_tap_packet_write_state *state_inout,

  /* Inputs. */
  const uint8_t *packet_buffer_in,
  bool packet_word_eop,
  uint16_t packet_word_length,
  const struct nanotube_tap_packet_write_req *req_in,
  const uint8_t *request_bytes_in,
  const uint8_t *request_mask_in);

///////////////////////////////////////////////////////////////////////////

/*! The core of the resize tap ingress stage.
**
** \param packet_word_length The number of bytes per packet word.
** This must be a constant expression for synthesis.
**
** \param packet_done_out A pointer to an output flag which will be
** set when the end of the packet has been processed.  When set, the
** next call should use the next request.
**
** \param cword_out A pointer to an output buffer which is written
** with the control word for this packet word.  The control word and
** packet word must be presented together to the egress stage.
**
** \param packet_length_out A pointer to a value that is set to the new 
** length of the packet (after the resize operation completes).  
**
** \param state The ingress stage state which is used to track
** progress through the packet.
**
** \param resize_req_in A pointer to an input buffer which contains
** the resize request for this packet.
**
** \param word_length_in The number of bytes in the current packet
** word.
**
** \param word_eop_in A flag indicating that the current word is the
** last word of the packet.
**
** This is the inner core of the resize tap ingress stage.  It is
** expected that users will use a wrapper for the appropriate bus type
** instead of calling this function directly.
**/
void nanotube_tap_packet_resize_ingress_core(
  /* Parameters. */
  nanotube_tap_offset_t packet_word_length,

  /* Outputs. */
  bool *packet_done_out,
  nanotube_tap_packet_resize_cword_t *cword_out,
  nanotube_tap_offset_t *packet_length_out,

  /* State. */
  nanotube_tap_packet_resize_ingress_state_t *state,

  /* Inputs. */
  nanotube_tap_packet_resize_req_t *resize_req_in,
  nanotube_tap_offset_t word_length_in,
  bool word_eop_in);


/*! The core of the resize tap egress stage.
**
** \param packet_word_length The number of bytes per packet word.
**  This must be a constant expression for synthesis.
**
** \param packet_word_index_bits The number of bits required for an
**  index into the packet buffer.  This must be a constant expression
**  for synthesis.
**
** \param input_done_out A pointer to an output flag which will be set
**  when the packet word has been processed.  When set, the next call
**  should use the next packet word and control word.
**
** \param word_valid_out A pointer to an output flag which will be set
**  if the stage has produced an output packet word.
**
** \param word_eop_out A pointer to an output flag which will be set
**  if the output packet word is the last word of the packet.  This
**  flag is only valid of word_valid_out is set.
**
** \param word_length_out A pointer to an output value which will
**  contain the number of bytes in the output packet word.  This will
**  be equal to packet_word_length for every word except the last.
**  This value is only valid of word_valid_out is set.
**
** \param word_data_out A pointer to an output buffer which will
** contain the data bytes of the output packet word.  The buffer is
** only valid of word_valid_out is set.
**
** \param state A pointer to the state buffer for the egress stage.
** This buffer is used to track progress through the packet.
**
** \param carried_data The bytes carried from the previous call to the
**  next call. This buffer must live across invocations of the tap, and
**  contain at least packet_word_len bytes.
**
** \param cword A pointer to an input buffer which contains the
** control word from the ingress stage.
**
** \param packet_data_in A pointer to an input bufer which contains
** the data bytes of the input packet word.
**
** This is the inner core of the resize tap ingress stage.  It is
** expected that users will use a wrapper for the appropriate bus type
** instead of calling this function directly.
*/
void nanotube_tap_packet_resize_egress_core(
  /* Parameters. */
  nanotube_tap_offset_t packet_word_length,
  unsigned int packet_word_index_bits,

  /* Outputs. */
  bool *input_done_out,
  bool *word_valid_out,
  bool *word_eop_out,
  nanotube_tap_offset_t *word_length_out,
  uint8_t *word_data_out,

  /* State. */
  nanotube_tap_packet_resize_egress_state_t *state,
  uint8_t *carried_data,

  /* Inputs. */
  nanotube_tap_packet_resize_cword_t *cword,
  uint8_t *packet_data_in);


/**@}*/

///////////////////////////////////////////////////////////////////////////

#ifdef __cplusplus
}
#endif

#endif // NANOTUBE_PACKET_TAPS_CORE_H
