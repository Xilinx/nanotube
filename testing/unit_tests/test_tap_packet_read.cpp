/**************************************************************************\
*//*! \file test_tap_packet_read.cpp
** \author  Neil Turton <neilt@amd.com>
**  \brief  A simple test for the packet read tap core logic.
**   \date  2020-09-29
*//*
\**************************************************************************/

/**************************************************************************
** Copyright (C) 2023, Advanced Micro Devices, Inc. All rights reserved.
** SPDX-License-Identifier: MIT
**************************************************************************/
#include "nanotube_packet_taps_core.h"
#include "test.hpp"

#include <algorithm>
#include <cassert>
#include <cstdlib>
#include <cstring>
#include <iomanip>
#include <iostream>

///////////////////////////////////////////////////////////////////////////

void test_tap_packet_read_core_case(
  struct nanotube_tap_packet_read_state *state,
  uint16_t result_buffer_length,
  uint8_t result_buffer_index_bits,
  uint16_t packet_buffer_length,
  uint8_t packet_buffer_index_bits,
  uint16_t read_offset,
  uint16_t read_length,
  uint16_t req_valid_offset,
  uint16_t packet_length)
{
  // The amount of padding to make sure there is no buffer over-run.
  uint16_t pad = 8;

  // The lengths of the buffers passed to the packet tap.
  uint16_t rb_len = result_buffer_length;
  uint16_t pb_len = packet_buffer_length;

  // The word index at which req.valid is set.
  uint16_t req_valid_index = ( req_valid_offset / pb_len );

  // The index of the word containing the first byte read.
  uint16_t read_start_index = ( read_offset / pb_len );

  // The byte after the last one read.
  uint16_t read_end_offset = (read_offset + read_length);

  // The offset into the packet at which the read completes.
  uint16_t resp_valid_offset =
    std::min(read_length == 0 ? req_valid_offset : read_end_offset-1U,
             packet_length-1U);

  // The word index at which the read completes and resp.valid should
  // be set.
  uint16_t resp_valid_index = (resp_valid_offset / pb_len);

  // The expected number of bytes read.
  uint16_t resp_length = ( std::min(read_end_offset, packet_length) -
                           std::min(read_offset, packet_length) );

  // The number of the packet words required.
  uint16_t packet_num_words = ( (packet_length + (pb_len-1)) / pb_len );

  // The number of bytes required to fill all the packet words.  This
  // can be larger than the packet length since this value includes
  // the padding bytes in the final word.
  uint16_t pd_len = packet_num_words * pb_len;

  if (test_verbose) {
    std::cout << "  Parameters: rbl=" << result_buffer_length
              << " pbl=" << packet_buffer_length
              << " ro=" << read_offset
              << " rl=" << read_length
              << " rvo=" << req_valid_offset
              << " pl=" << packet_length
              << "\n";
    std::cout << "  Expecting: rvo=" << uint32_t(resp_valid_offset)
              << ", rvi=" << uint32_t(resp_valid_index)
              << ", rl=" << uint32_t(resp_length)
              << "\n";
  }

  assert(req_valid_index <= read_start_index);
  assert(resp_valid_index < packet_num_words);

  // Variables to pass to the packet tap.
  struct nanotube_tap_packet_read_resp resp;
  uint8_t result_buffer[rb_len+2*pad];
  uint8_t packet_buffer[pb_len+2*pad];
  struct nanotube_tap_packet_read_req req;

  // An array of zeros for checking that the result buffer is
  // correctly padded with zeros.
  uint8_t zeros[rb_len];
  memset(zeros, 0, sizeof(zeros));

  // An array of random data for initializing input buffers.
  uint8_t rand_data[(rb_len+2*pad) + (pd_len+2*pad)];
  const uint8_t *rd_rb = rand_data + 0;
  const uint8_t *rd_pd = rand_data + (rb_len+2*pad);

  for (size_t i = 0; i < sizeof(rand_data); i++)
    rand_data[i] = rand() & 0xff;

  // Fill the result buffer with random data to make sure everything
  // gets written.
  memcpy(result_buffer, rd_rb, sizeof(result_buffer));

  for (uint16_t index=0; index<packet_num_words; index++) {
    // Fill in the request, setting the valid bit at the requested
    // time.
    if (index >= req_valid_index) {
      req.valid = 1;
      req.read_offset = read_offset;
      req.read_length = read_length;
    } else {
      req.valid = 0;
      req.read_offset = rand() & 0xffff;
      req.read_length = rand() & 0xffff;
    }

    // Determine the packet arguments provided to the packet tap.
    bool packet_word_eop = (index == packet_num_words-1);
    uint16_t packet_word_length = pb_len;
    if (packet_word_eop)
      packet_word_length = packet_length - (packet_num_words-1)*pb_len;

    // Initialise the packet buffer provided to the packet tap.
    memcpy(packet_buffer, rd_pd, pad);
    memcpy(packet_buffer+pad, rd_pd + pad + index*pb_len, pb_len);
    memcpy(packet_buffer+pad+pb_len, rd_pd + pad + pd_len, pad);

    if (test_verbose) {
      std::cout << "  Invoking tap: i=" << uint16_t(index)
                << " rbl=" << uint32_t(result_buffer_length)
                << " rbib=" << uint32_t(result_buffer_index_bits)
                << " pbl=" << uint32_t(packet_buffer_length)
                << " pbib=" << uint32_t(packet_buffer_index_bits)
                << " eop=" << uint32_t(packet_word_eop)
                << " pwl=" << uint32_t(packet_word_length)
                << " rv=" << uint32_t(req.valid)
                << " ro=" << uint32_t(req.read_offset)
                << " rl=" << uint32_t(req.read_length)
                << "\n";
    }

    // Call the packet tap.
    nanotube_tap_packet_read_core(
      result_buffer_length,
      result_buffer_index_bits,
      packet_buffer_length,
      packet_buffer_index_bits,
      &resp, result_buffer+pad,
      state,
      packet_buffer+pad, packet_word_eop, packet_word_length, &req);

    if (test_verbose) {
      std::cout << "  Response: rv=" << uint32_t(resp.valid);
      if (resp.valid)
        std::cout << " rl=" << uint32_t(resp.result_length);
      std::cout << "\n";
    }

    // Check the arrays which should not be modified.
    assert_array_eq(packet_buffer, rd_pd, pad);
    assert_array_eq(packet_buffer+pad, rd_pd + pad + index*pb_len, pb_len);
    assert_array_eq(packet_buffer+pad+pb_len, rd_pd + pad + pd_len, pad);
    assert_array_eq(result_buffer, rd_rb, pad);
    assert_array_eq(result_buffer+result_buffer_length+pad,
                    rd_rb+result_buffer_length+pad, pad);

    // Make sure the valid bit is set at the right time.
    assert_eq(resp.valid, index==resp_valid_index);

    // Check the response if it was produced and should have been
    // produced.
    if (resp.valid && index==resp_valid_index) {
      // Check the response length.
      assert_eq(resp.result_length, resp_length);

      // Check the data copied from the packet.
      if (resp_length != 0) {
        uint16_t data_length = std::min(resp_length, rb_len);
        assert(read_offset < packet_length);
        assert(resp_length <= packet_length - read_offset);
        assert_array_eq(result_buffer+pad, rd_pd+pad+read_offset,
                        data_length);
      }

      // Check the zero padding.
      if (resp_length < rb_len)
        assert_array_eq(result_buffer+pad+resp_length, zeros,
                        rb_len - resp_length);
    }
  }
}

void test_tap_packet_read_core()
{
  struct nanotube_tap_packet_read_state state;

  // Case description codes indicate the use of each packet byte
  // assuming a 4 byte word size.  The actual buffer size varies so
  // the description code only shows the kind of case being tested,
  // indicating whether the read begins at a word boundary etc...
  //
  // The codes are:
  //   Be - The byte is before the read.
  //   Rf - The byte is the full read.
  //   Rs - The byte is the start of the read.
  //   Rm - The byte is mid-read.
  //   Re - The byte is the end of the read.
  //   Af - The byte is after the read.

  std::cout << "Case  1: Rf Af Af Af\n";

  state = nanotube_tap_packet_read_state_init;

  test_tap_packet_read_core_case(&state,
                                 /* Result buffer */   8, 3,
                                 /* Packet buffer */   8, 3,
                                 /* Read offset/len */ 0, 1,
                                 /* Req valid */       0,
                                 /* Packet length */   8);

  std::cout << "Case  2: Rs Rm Re Af\n";

  test_tap_packet_read_core_case(&state,
                                 /* Result buffer */   7, 3,
                                 /* Packet buffer */   7, 3,
                                 /* Read offset/len */ 7, 4,
                                 /* Req valid */       0,
                                 /* Packet length */   17);

  std::cout << "Case  3: Rs Rm Rm Re\n";

  test_tap_packet_read_core_case(&state,
                                 /* Result buffer */   29, 5,
                                 /* Packet buffer */   9, 4,
                                 /* Read offset/len */ 9, 9,
                                 /* Req valid */       9,
                                 /* Packet length */   18);

  std::cout << "Case  4: Rs Rm Rm Rm | Rm Re Af Af\n";

  test_tap_packet_read_core_case(&state,
                                 /* Result buffer */   35, 6,
                                 /* Packet buffer */   13, 4,
                                 /* Read offset/len */ 0, 21,
                                 /* Req valid */       0,
                                 /* Packet length */   96);

  std::cout << "Case  5: Rs Rm Rm Rm | Rm Rm Rm Re\n";

  test_tap_packet_read_core_case(&state,
                                 /* Result buffer */   31, 6,
                                 /* Packet buffer */   13, 4,
                                 /* Read offset/len */ 0, 26,
                                 /* Req valid */       0,
                                 /* Packet length */   96);

  std::cout << "Case  6: Rs Rm Rm Rm | Rm Rm Rm Rm | Re Af Af Af\n";

  test_tap_packet_read_core_case(&state,
                                 /* Result buffer */   43, 6,
                                 /* Packet buffer */   15, 4,
                                 /* Read offset/len */ 0, 37,
                                 /* Req valid */       0,
                                 /* Packet length */   73);

  std::cout << "Case  7: Be Rf Af Af\n";

  test_tap_packet_read_core_case(&state,
                                 /* Result buffer */   35, 6,
                                 /* Packet buffer */   13, 4,
                                 /* Read offset/len */ 0, 31,
                                 /* Req valid */       0,
                                 /* Packet length */   73);

  std::cout << "Case  8: Be Rs Re Af\n";

  test_tap_packet_read_core_case(&state,
                                 /* Result buffer */   27, 5,
                                 /* Packet buffer */   11, 4,
                                 /* Read offset/len */ 5, 4,
                                 /* Req valid */       0,
                                 /* Packet length */   30);

  std::cout << "Case  9: Be Rs Rm Re\n";

  test_tap_packet_read_core_case(&state,
                                 /* Result buffer */   29, 5,
                                 /* Packet buffer */   11, 4,
                                 /* Read offset/len */ 16, 6,
                                 /* Req valid */       16,
                                 /* Packet length */   26);

  std::cout << "Case 10: Be Rs Rm Rm | Rm Re Af Af\n";

  test_tap_packet_read_core_case(&state,
                                 /* Result buffer */   29, 5,
                                 /* Packet buffer */   12, 4,
                                 /* Read offset/len */ 8, 6,
                                 /* Req valid */       0,
                                 /* Packet length */   20);

  std::cout << "Case 11: Rs Rm <EOP>\n";

  test_tap_packet_read_core_case(&state,
                                 /* Result buffer */   31, 5,
                                 /* Packet buffer */   18, 5,
                                 /* Read offset/len */ 0, 12,
                                 /* Req valid */       0,
                                 /* Packet length */   10);

  std::cout << "Case 12: Rs Rm Rm Rm <EOP>\n";

  test_tap_packet_read_core_case(&state,
                                 /* Result buffer */   31, 5,
                                 /* Packet buffer */   17, 5,
                                 /* Read offset/len */ 0, 12,
                                 /* Req valid */       0,
                                 /* Packet length */   10);

  std::cout << "Case 13: Rs Rm Rm Rm | Rm <EOP>\n";

  test_tap_packet_read_core_case(&state,
                                 /* Result buffer */   26, 5,
                                 /* Packet buffer */   19, 5,
                                 /* Read offset/len */ 19, 26,
                                 /* Req valid */       0,
                                 /* Packet length */   40);

  std::cout << "Case 14: Be Rs Rm <EOP>\n";

  test_tap_packet_read_core_case(&state,
                                 /* Result buffer */   26, 5,
                                 /* Packet buffer */   21, 5,
                                 /* Read offset/len */ 11, 26,
                                 /* Req valid */       0,
                                 /* Packet length */   19);

  std::cout << "Case 15: Be Rs Rm Rm <EOP>\n";

  test_tap_packet_read_core_case(&state,
                                 /* Result buffer */   24, 5,
                                 /* Packet buffer */   21, 5,
                                 /* Read offset/len */ 16, 23,
                                 /* Req valid */       0,
                                 /* Packet length */   21);

  std::cout << "Case 16: Be Rs Rm Rm | Rm <EOP>\n";

  test_tap_packet_read_core_case(&state,
                                 /* Result buffer */   25, 5,
                                 /* Packet buffer */   29, 5,
                                 /* Read offset/len */ 26, 20,
                                 /* Req valid */       0,
                                 /* Packet length */   32);

  std::cout << "Random cases.\n";
  for (int i=0; i<1000; i++) {
    uint16_t result_buffer_length = (rand() % 16) + 1;
    uint16_t packet_buffer_length = (rand() % 16) + 1;
    uint16_t read_offset = (rand() % 32);
    uint16_t read_length = (rand() % 16);
    uint16_t req_valid_offset = (rand() % (read_offset+1));
    uint16_t packet_length = (rand() % 32) + 1;

    if (test_verbose)
      std::cout << "Random case " << i << "\n";

    test_tap_packet_read_core_case(&state,
                                   result_buffer_length, 6,
                                   packet_buffer_length, 6,
                                   read_offset, read_length,
                                   req_valid_offset,
                                   packet_length);
  }
}

int main(int argc, char *argv[])
{
  test_init(argc, argv);
  test_tap_packet_read_core();
  return test_fini();
}

///////////////////////////////////////////////////////////////////////////
