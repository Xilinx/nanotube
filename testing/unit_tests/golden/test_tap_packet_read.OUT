Case  1: Rf Af Af Af
Case  2: Rs Rm Re Af
Case  3: Rs Rm Rm Re
Case  4: Rs Rm Rm Rm | Rm Re Af Af
Case  5: Rs Rm Rm Rm | Rm Rm Rm Re
Case  6: Rs Rm Rm Rm | Rm Rm Rm Rm | Re Af Af Af
Case  7: Be Rf Af Af
Case  8: Be Rs Re Af
Case  9: Be Rs Rm Re
Case 10: Be Rs Rm Rm | Rm Re Af Af
Case 11: Rs Rm <EOP>
Case 12: Rs Rm Rm Rm <EOP>
Case 13: Rs Rm Rm Rm | Rm <EOP>
Case 14: Be Rs Rm <EOP>
Case 15: Be Rs Rm Rm <EOP>
Case 16: Be Rs Rm Rm | Rm <EOP>
Random cases.
Test passed.
